src/Tools/dotnet-monitor/CollectionRules/Configuration/CollectionRulePostConfigureNamedOptions.cs

```diff
@@ -0,0 +1,147 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Diagnostics.Tools.Monitor.CollectionRules.Options;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.Diagnostics.Tools.Monitor.CollectionRules.Configuration
+{
+    internal sealed class CollectionRulePostConfigureOptions :
+        IPostConfigureOptions<CollectionRuleOptions>
+    {
+        private readonly IOptionsMonitor<CustomShortcutOptions> _shortcutOptions;
+        private readonly IConfiguration _configuration;
+        private readonly ILogger<CollectionRulePostConfigureOptions> _logger;
+
+        public CollectionRulePostConfigureOptions(
+            IOptionsMonitor<CustomShortcutOptions> shortcutOptions,
+            IConfiguration configuration,
+            ILogger<CollectionRulePostConfigureOptions> logger)
+        {
+            _shortcutOptions = shortcutOptions;
+            _configuration = configuration;
+            _logger = logger;
+        }
+
+        public void PostConfigure(string name, CollectionRuleOptions options)
```

Comment:
This is doing some complicated math and relying on configuration binding to not bind if the trigger/action/etc is not fully specified. This can be simplified by doing something like:
- In `CollectionRuleConfigureNamedOptions`, remove the `BindTriggerSettings` and `BindActionSettings` methods/calls and move them to this class. That class will only bind the `CollectionRuleOptions` instance to the collection rule configuration section.
- Instead of calculating where something was fulfilled or not, just clear each of the lists on `CollectionRuleOptions`. For each child section in each list section, either (1) if the section doesn't have a body, then assign the body from corresponding shortcut options, or (2) bind and call `BindXSettings` method as usual.

There's no math involved in determining the order of items or where to insert new items.

Reply:
Without preserving its contents, are we sure we could actually clear out the lists on `CollectionRuleOptions`? If we throw away the `CollectionRuleActionOptions`, when we try to `BindActionSettings` it'll try to get `actionOptions.Type` on the current action - which, if we throw them all away and just fill them in as we go, means we wouldn't have a type here. Or am I misunderstanding this part?

Reply:
You can clear it out and rebind it when necessary. The existing `CollectionRuleConfigureNamedOptions` already knows how to bind the `CollectionRuleOptions` without any help; the `CollectionRulePostConfigureNamedOptions` has the same context, so it can recreate it.

The code would look something like:
```cs
options.Actions.Clear();

IConfigurationSection actionsSection = ruleSection.GetSection(nameof(CollectionRuleOptions.Actions));
if (actionsSection.Exists())
{
    foreach (IConfigurationSection actionSection in actionsSection.GetChildren())
    {
        // actionSection.Key is the index of the action e.g. "0", "1", "2"
        // actionSection.Value is null (if specified inline) or non-empty string if a simple name
        if (string.IsNullOrEmpty(actionSection.Value))
        {
            // Bind action options from inline configuration
            CollectionRuleActionOptions actionOptions = new();

            actionSection.Bind(actionOptions);

            BindActionSettings(actionSection, actionOptions);

            options.Actions.Add(actionOptions);
        }
        else
        {
            // Get action options from template
            // options.Actions.Add(actionOptionsFromTemplate);
        }
    }
}

...

private void BindActionSettings(IConfigurationSection actionSection, CollectionRuleActionOptions actionOptions)
{
    if (null != actionOptions &&
        _actionOperations.TryCreateOptions(actionOptions.Type, out object actionSettings))
    {
        IConfigurationSection settingsSection = actionSection.GetSection(nameof(CollectionRuleActionOptions.Settings));

        settingsSection.Bind(actionSettings);

        actionOptions.Settings = actionSettings;
    }
}

```

Reply:
Gotcha, that makes sense - I had been saving a copy of the list as a queue before clearing it, and then just dequeuing instead of creating a new `CollectionRuleActionOptions` and binding it to `actionSection`, but I'll switch to this approach.